### PR TITLE
[Core] Investigate memory leaks, manual GC on ResetToGenesis - (Issue #440)

### LINKED ISSUE #440: [Core] Investigate memory leaks
## Objective

Make sure that we are not allocating memory without ever releasing it and other similar bugs that can lead to memory exhaustion.

## Origin Document

While investigating some instability in `LocalNet`, I have noticed that  some of the nodes were being killed for excessive memory pressure.

image

There might be several places where we should address this issue but as a starting point I have noticed that resetting to genesis via the debug CLI causes unwanted allocations.

In the video below, I focus on one of the nodes while I run multiple `ResetToGenesis` commands. Notice how the memory (top-right cell) increases in steps and then plateaus:

https://user-images.githubusercontent.com/29378614/212944742-a495685a-e4e2-4dd1-a156-79d760c80b90.mp4

## Goals

- Identify the root cause of the memory leaks
- Fix the memory leaks
- Identify if there are other areas of concerns in the codebase related to memory leaks and create follow-up tickets (e.g. an applied pattern that is a root cause that may present itself elsewhere in the codebase)

## Deliverable

- [ ] A working PR where multiple `ResetToGenesis` triggered sequentially do not cause node crashes anymore. 
- [ ] If other unrelated issues are identified along the way, please log them accordingly

## Non-goals / Non-deliverables

- Changes in the logic 

## General issue deliverables

- [ ] Update the appropriate CHANGELOG(s)
- [ ] Update any relevant local/global README(s)
- [ ] Update relevant source code tree explanations
- [ ] Add or update any relevant or supporting [mermaid](https://mermaid-js.github.io/mermaid/) diagrams

## Testing Methodology

- [ ] **Task specific tests or benchmarks**: `make ...`
- [ ] **New tests or benchmarks**: `make ...`
- [ ] **All tests**: `make test_all`
- [ ] **LocalNet**: verify a `LocalNet` is still functioning correctly by following the instructions at [docs/development/README.md](https://github.com/pokt-network/pocket/tree/main/docs/development)

---

**Creator**: @deblasis 
**Co-Owners**: @Olshansk, @gokutheengineer

### PR DESCRIPTION
**Pull request by deblasis**
## Description

This PR reclaims memory manually after the memory-churn intensive operations performed during a `ResetToGenesis` debug command

In particular:
```go
	// Clear the BlockStore
	if err := m.blockStore.ClearAll(); err != nil {
		return err
	}

	// Clear all the Trees
	if err := postgresCtx.clearAllTreeState(); err != nil {
		return err
	}
```

This is the result of my initial investigation https://github.com/pokt-network/pocket/issues/440#issuecomment-1387107092

## Issue

This might fix #440 but I'd like another pair of eyes on this to understand if this is the right approach

## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [x] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- Calling `debug.FreeOSMemory()` after cleaning up the persistence structures

## Testing

- [x] `make develop_test`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist
- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

File: persistence/docs/CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [0.0.0.23] - 2023-01-18
 
 - Consolidate the `Block` proto structures under `shared/core/types`
+- Added `debug.FreeOSMemory()` on `ResetToGenesis` to free-up memory and stabilize `LocalNet`.
```
Comment (Olshansk) [inline persistence/docs/CHANGELOG.md:13]: Prob need to update this now

**Review by Olshansk — APPROVED** [review]

**Comment by deblasis** [comment]
@Olshansk PTAL take a look at the short documentation I have added because of the comments in #440 

If it's good enough I'll merge 🙏

File: persistence/docs/CHANGELOG.md [hunk 2]
```diff
@@ -18,6 +22,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [0.0.0.23] - 2023-01-18
 
+- Consolidate the `Block` proto structures under `shared/core/types`
```
Comment (Olshansk) [inline persistence/docs/CHANGELOG.md:25]:
```suggestion
```

**Review by Olshansk — APPROVED** [review]
Awesome documentation, ty!

**Pull request merged**